src/development/coding-logging.md

```diff
@@ -0,0 +1,251 @@
+  * On failed input validation (although WARN may be used instead)
+* `WARN` is used to present undesirable but non-critical errors in application flow
+  For example:
+  * On failed authentication
+  * On failed authorisation
+  * On client error
+  * On server error that is recoverable
+  * On session management failures
+  Events logged at WARN are candidates for monitoring so that alerts may be raised after multiple events
+* `ERROR` is used to present errors in application flow that require immediate attention
+  For example:
+  * On server error that is unrecoverable
+  * On failed output validation (e.g. database failed to insert)
+  * On failed application startup/shutdown
+  All events logged at ERROR should be monitored so that alerts may be raised after a single event
+
+## Meta information
+
+The following meta information should be logged:
+
+* unique ID per log statement
+* timestamp in ISO8601 format with minimum milliseconds precision and timezone information
+* unique host identifier
+* containerised environments should uniquely identify the container
+* application/system name
+* application specific _category_ e.g. class name
+* process ID
+* thread ID
+* unique request ID
+* unique correlation ID for requests that span multiple applications and services
+
+The following meta information must not be logged:
+
+* client session ID
+  Client held session IDs should not be logged, as this opens a security risk.
+  If tracing session interactions is required, then it is permissable to log a secure hash of the session ID.
+
+## Logging Guidance
+
+Never print logging information direct to 'system output' except through configuration of a logging library. For instance, avoid statements like this:
+
+```java [g1:Java]
+System.out.println("A useful log statement");
+```
+
+```javascript [g1:Javascript]
+console.log("A useful log statement");
+```
+
+Logging libraries should provide data as arguments to log messages with placeholders. This has performance benefits in languages such as Java, but more importantly it allows data to be sanitised within the library. For instance, use this form:
+
+```java [g1:Java]
+logger.info("Here is some data - {}", data);
+```
+
+```javascript [g1:Javascript]
+logger.log('info', 'Here is some data -  %s', data);
+```
+
+and avoid this form:
+
+```java [g1:Java]
+logger.info("Here is some data - " + data);
+```
+
+```javascript [g1:Javascript]
+logger.log('info', 'Here is some data - ' + data);
+```
+
+### Content
+
+Log entries should represent atomic events with enough contextual information to understand that event.
+
+* It should not be necessary to cross reference one log entry with another to understand it.
+* Contextual information should include the application flow and logic that caused the event
+* Contextual information should include enough data to identify what caused the event
+
+```java [g1:Java]
+logger.info("User {} successfully authenticated", user.getId())
```

Comment:
Can find a good reference right now, but following the convention:

`logger.info("User: {} successfully authenticated", user.getId())`

Will allow Datadog to 'pickout' the values in the log